src/par_iter/internal.rs

```diff
@@ -83,6 +83,13 @@ pub trait UnindexedConsumer<ITEM>: Consumer<ITEM> {
     fn to_reducer(&self) -> Self::Reducer;
 }
 
+/// An unindexed producer that doesn't know its exact length.
+/// (or can't represent its known length in a `usize`)
+pub trait UnindexedProducer: IntoIterator + Send + Sized {
```

Comment:
Great! One nit: I've always imagined that we would have `Producer: UnindexedProducer`; but I guess it's just kind of a lot of boilerplate.

I would think something like this:

```rust
impl<T: Producer> UnindexedProducer for T {
    fn can_split(&self) -> bool { self.len() != 0 }
    fn split(&self) -> (Self, Self) { self.split_at(self.len() / 2) }
}
```

But I think we'd have to add the `len` method to `Producer` -- not that there is a good reason we don't have it.

Mostly I like the symmetry where you have:

- `Consumer` and `UnindexedConsumer: Consumed`
- `UnindexedProducer` and `Producer: UnindexedProducer`